src/pkey/clu_pkey.c

```diff
@@ -0,0 +1,257 @@
+                        ret = BAD_FUNC_ARG;
+                    }
+                }
+
+                if (ret == 0) {
+                     der = malloc(derSz);
+                     if (der == NULL) {
+                         printf("unable to malloc der buffer\n");
+                         ret = MEMORY_E;
+                     }
+                }
+
+                if (ret == 0) {
+                     ret = wc_EccPublicKeyToDer((ecc_key*)ec->internal, der,
+                             derSz, 1);
+                     if (ret > 0) {
+                         ret    = 0;
+                         *out   = der;
+                         *outSz = derSz;
+                     }
+                     else {
+                        ret = BAD_FUNC_ARG;
+                        printf("decoding der from internal structure failed\n");
+                     }
+                }
+
+                if (der != NULL)
+                    free(der);
+            }
+            break;
+
+        default:
+            printf("unknown / unsupported key type\n");
+            ret = USER_INPUT_ERROR;
+    }
+
+    if (ret == 0)
+        ret = *outSz;
+    return ret;
+}
+
+
+int wolfCLU_pKeySetup(int argc, char** argv)
+{
+    int ret    = 0;
+    int inForm = PEM_FORM;
+    int pubOut = 0;
+    int option;
+    int longIndex = 1;
+    WOLFSSL_EVP_PKEY *pkey = NULL;
+    WOLFSSL_BIO *bioIn  = NULL;
+    WOLFSSL_BIO *bioOut = NULL;
+
+    opterr = 0; /* do not display unrecognized options */
+    optind = 0; /* start at indent 0 */
+    while ((option = getopt_long_only(argc, argv, "",
+                   pkey_options, &longIndex )) != -1) {
+        switch (option) {
+            case WOLFCLU_PUBOUT:
+                pubOut = 1;
+                break;
+
+            case WOLFCLU_INFILE:
+                bioIn = wolfSSL_BIO_new_file(optarg, "rb");
+                if (bioIn == NULL) {
+                    printf("unable to open public key file %s\n", optarg);
+                    ret = -1;
+                }
+                break;
+
+            case WOLFCLU_INFORM:
+                inForm = wolfCLU_checkInform(optarg);
+                break;
+
+            case WOLFCLU_HELP:
+                wolfCLU_pKeyHelp();
+                return 0;
+
+            case ':':
+            case '?':
+                break;
+
+            default:
+                /* do nothing. */
+                (void)ret;
+        }
+    }
+
+
+    if (ret == 0 && bioIn != NULL) {
+        if (inForm == PEM_FORM) {
+            pkey = wolfSSL_PEM_read_bio_PrivateKey(bioIn, NULL, NULL, NULL);
+        }
+        else {
+            pkey = wolfSSL_d2i_PrivateKey_bio(bioIn, NULL);
+        }
+        if (pkey == NULL) {
+            printf("error reading key from file\n");
+            ret = USER_INPUT_ERROR;
+        }
+    }
+
+    if (ret == 0) {
+        bioOut = wolfSSL_BIO_new(wolfSSL_BIO_s_file());
+        if (bioOut == NULL) {
+            ret = -1;
```

Comment:
Future improvement: perhaps -1 could become a named error. It seems we're using it as a catch-all for general errors, so maybe `GENERAL_ERROR`?